File: runtime-elixir/client/smithy_retry.ex
defmodule SmithyRetry do
  @moduledoc """
  Retry logic for Smithy-generated Elixir clients.

  Ported from `aws_retry.erl`. Implements exponential back-off with optional
  jitter for retrying transient AWS errors.
  """

  @default_max_retries 3
  @default_initial_backoff_ms 100
  @default_max_backoff_ms 20_000
  @default_multiplier 2.0
  @default_use_jitter true

  @doc """
  Execute `fun` and retry on retryable errors.

  ## Options

    * `:max_retries` — number of additional attempts (default: `#{@default_max_retries}`)
    * `:initial_backoff_ms` — first backoff in ms (default: `#{@default_initial_backoff_ms}`)
    * `:max_backoff_ms` — ceiling for backoff (default: `#{@default_max_backoff_ms}`)
    * `:multiplier` — exponential factor (default: `#{@default_multiplier}`)
    * `:use_jitter` — add random jitter (default: `#{@default_use_jitter}`)
    * `:enable_retry` — set to `false` to disable all retrying (default: `true`)
    * `:logger` — `fun/1` called with a retry-event map on each retry
  """
  @spec with_retry((() -> {:ok, term()} | {:error, term()}), keyword()) ::
          {:ok, term()} | {:error, term()}
  def with_retry(fun, opts \\ []) when is_function(fun, 0) do
    unless Keyword.get(opts, :enable_retry, true) do
      fun.()
    else
      max_retries      = Keyword.get(opts, :max_retries,       @default_max_retries)
      initial_backoff  = Keyword.get(opts, :initial_backoff_ms, @default_initial_backoff_ms)
      max_backoff      = Keyword.get(opts, :max_backoff_ms,     @default_max_backoff_ms)
      multiplier       = Keyword.get(opts, :multiplier,         @default_multiplier)
      use_jitter       = Keyword.get(opts, :use_jitter,         @default_use_jitter)
      logger           = Keyword.get(opts, :logger)

      retry_loop(fun, 0, max_retries, initial_backoff, max_backoff, multiplier, use_jitter, logger)
    end
  end

  @doc "Return true if the error is a transient/retryable AWS error."
  @spec retryable?(term()) :: boolean()
  def retryable?({:http_error, status, _}) when status in [429, 500, 502, 503, 504], do: true
  def retryable?({:http_error, _, _}), do: false
  def retryable?(:timeout), do: true
  def retryable?(:connection_refused), do: true
  def retryable?({:aws_error, _, code, _}) when code in ["ThrottlingException",
                                                          "RequestThrottled",
                                                          "ServiceUnavailable",
                                                          "InternalError",
                                                          "InternalFailure"], do: true
  def retryable?(_), do: false

  @doc """
  Polls `poll` repeatedly with exponential backoff until it reports terminal
  success or failure, or the maximum number of attempts is reached.

  This is the polling primitive used by generated Smithy waiters
  (`@waitable` trait). The poller is the generated acceptor evaluator which
  calls the underlying operation and matches the result against the waiter's
  acceptors.

  `poll` must return one of:

    * `{:success, value}` — terminal success; `wait/3` returns `{:ok, value}`.
    * `{:failure, reason}` — terminal failure; `wait/3` returns `{:error, reason}`.
    * `{:retry, value}` — sleep with exponential backoff and call `poll` again.

  `min_delay` and `max_delay` are seconds (matching the `@waitable` trait).
  After 100 retry attempts (hard cap, no setting today) `wait/3` gives up
  with `{:error, :max_attempts_exceeded}`.
  """
  @spec wait((() -> {:success | :failure | :retry, term()}), non_neg_integer(), non_neg_integer()) ::
          {:ok, term()} | {:error, term()}
  def wait(poll, min_delay, max_delay) when is_function(poll, 0) do
    wait_loop(poll, 0, 100, min_delay * 1000, max_delay * 1000)
  end

  # ---------------------------------------------------------------------------
  # Private helpers
  # ---------------------------------------------------------------------------

  defp wait_loop(_poll, attempt, max_attempts, _min_ms, _max_ms) when attempt >= max_attempts do
    {:error, :max_attempts_exceeded}
  end

  defp wait_loop(poll, attempt, max_attempts, min_ms, max_ms) do
    case poll.() do
      {:success, value} ->
        {:ok, value}

      {:failure, reason} ->
        {:error, reason}

      {:retry, _value} ->
        base_ms = round(min_ms * :math.pow(2, attempt))
        delay_ms = min(base_ms, max_ms)
        Process.sleep(delay_ms)
        wait_loop(poll, attempt + 1, max_attempts, min_ms, max_ms)
    end
  end

  defp retry_loop(fun, attempt, max_retries, initial_backoff, max_backoff, multiplier, use_jitter, logger) do
    case fun.() do
      {:ok, _} = ok ->
        ok

      {:error, reason} when attempt >= max_retries ->
        {:error, reason}

      {:error, reason} ->
        if retryable?(reason) do
          backoff =
            (initial_backoff * :math.pow(multiplier, attempt))
            |> round()
            |> min(max_backoff)
            |> maybe_jitter(use_jitter)

          if logger, do: logger.(%{
            event: :retry_attempt,
            attempt: attempt + 1,
            max_retries: max_retries,
            backoff_ms: backoff,
            error: reason
          })

          Process.sleep(backoff)

          retry_loop(fun, attempt + 1, max_retries, initial_backoff, max_backoff, multiplier, use_jitter, logger)
        else
          {:error, reason}
        end
    end
  end

  defp maybe_jitter(backoff, false), do: backoff
  defp maybe_jitter(backoff, true) do
    jitter = :rand.uniform(max(div(backoff, 4), 1))
    backoff + jitter
  end
end

File: runtime-erlang/client/smithy_retry.erl
-module(smithy_retry).

%% Retry logic with exponential backoff for AWS API calls
%% Handles transient errors (5xx, 429) with configurable retry strategy

-export([
    with_retry/1,
    with_retry/2,
    is_retryable_error/1,
    wait/3
]).

%%%===================================================================
%%% API Functions
%%%===================================================================

%% @doc Retry a function with default options
%% Default: 3 retries, 100ms initial backoff, 20s max backoff
-spec with_retry(Fun :: fun(() -> {ok, term()} | {error, term()})) ->
    {ok, term()} | {error, term()}.
with_retry(Fun) ->
    with_retry(Fun, #{}).

%% @doc Retry a function with custom options
%% Options:
%%   - max_retries: Maximum number of retry attempts (default: 3)
%%   - initial_backoff: Initial backoff in milliseconds (default: 100)
%%   - max_backoff: Maximum backoff in milliseconds (default: 20000)
%%   - backoff_multiplier: Multiplier for exponential backoff (default: 2)
%%   - jitter: Add jitter to backoff (default: true)
%%   - retryable_errors: Custom list of retryable error patterns (default: standard AWS errors)
-spec with_retry(
    Fun :: fun(() -> {ok, term()} | {error, term()}),
    Options :: map()
) ->
    {ok, term()} | {error, term()}.
with_retry(Fun, Options) when is_function(Fun, 0), is_map(Options) ->
    MaxRetries = maps:get(max_retries, Options, 3),
    InitialBackoff = maps:get(initial_backoff, Options, 100),
    MaxBackoff = maps:get(max_backoff, Options, 20000),
    Multiplier = maps:get(backoff_multiplier, Options, 2),
    UseJitter = maps:get(jitter, Options, true),

    retry_loop(Fun, 0, MaxRetries, InitialBackoff, MaxBackoff, Multiplier, UseJitter, Options).

%% @doc Check if an error is retryable
%% Retryable errors:
%%   - HTTP 5xx (server errors)
%%   - HTTP 429 (too many requests)
%%   - Connection errors: timeout, econnrefused, ehostunreach, etc.
-spec is_retryable_error(Error :: term()) -> boolean().
is_retryable_error({StatusCode, _}) when is_integer(StatusCode) ->
    StatusCode >= 500 orelse StatusCode =:= 429;
is_retryable_error({error, {StatusCode, _}}) when is_integer(StatusCode) ->
    StatusCode >= 500 orelse StatusCode =:= 429;
is_retryable_error({error, timeout}) ->
    true;
is_retryable_error({error, etimedout}) ->
    true;
is_retryable_error({error, econnrefused}) ->
    true;
is_retryable_error({error, ehostunreach}) ->
    true;
is_retryable_error({error, enetunreach}) ->
    true;
is_retryable_error({error, econnreset}) ->
    true;
is_retryable_error({error, {failed_connect, _}}) ->
    true;
is_retryable_error(_) ->
    false.

%% @doc Polls `Poll' repeatedly with exponential backoff until it reports
%% terminal success or failure, or the maximum number of attempts is reached.
%%
%% This is the polling primitive used by generated Smithy waiters
%% (`@waitable' trait). The poller is the generated acceptor evaluator
%% which calls the underlying operation and matches the result against the
%% waiter's acceptors.
%%
%% `Poll' must return one of:
%%   `{success, Value}' — terminal success; `wait/3' returns `{ok, Value}'.
%%   `{failure, Reason}' — terminal failure; `wait/3' returns `{error, Reason}'.
%%   `{retry,   Value}' — sleep with exponential backoff and call `Poll' again.
%%
%% `MinDelay' and `MaxDelay' are seconds (matching the `@waitable' trait).
%% After 100 retry attempts (hard cap, no setting today) `wait/3' gives up
%% with `{error, max_attempts_exceeded}'.
-spec wait(
    Poll :: fun(() -> {success | failure | retry, term()}),
    MinDelay :: non_neg_integer(),
    MaxDelay :: non_neg_integer()
) ->
    {ok, term()} | {error, term()}.
wait(Poll, MinDelay, MaxDelay) when is_function(Poll, 0) ->
    wait_loop(Poll, 0, 100, MinDelay * 1000, MaxDelay * 1000).

%%%===================================================================
%%% Internal Functions
%%%===================================================================

%% @private
%% Waiter polling loop with exponential backoff.
wait_loop(_Poll, Attempt, MaxAttempts, _MinMs, _MaxMs) when Attempt >= MaxAttempts ->
    {error, max_attempts_exceeded};
wait_loop(Poll, Attempt, MaxAttempts, MinMs, MaxMs) ->
    case Poll() of
        {success, Value} ->
            {ok, Value};
        {failure, Reason} ->
            {error, Reason};
        {retry, _Value} ->
            BaseMs = MinMs * round(math:pow(2, Attempt)),
            DelayMs = min(BaseMs, MaxMs),
            timer:sleep(DelayMs),
            wait_loop(Poll, Attempt + 1, MaxAttempts, MinMs, MaxMs)
    end.

%% @private
%% Main retry loop with exponential backoff
retry_loop(
    Fun, Attempt, MaxRetries, _InitialBackoff, _MaxBackoff, _Multiplier, _UseJitter, _Options
) when
    Attempt >= MaxRetries
->
    %% Final attempt - no more retries after this
    case Fun() of
        {ok, Result} ->
            {ok, Result};
        {error, Reason} ->
            %% Return error with retry information
            {error,
                {max_retries_exceeded, #{
                    attempts => Attempt + 1,
                    last_error => Reason
                }}}
    end;
retry_loop(Fun, Attempt, MaxRetries, InitialBackoff, MaxBackoff, Multiplier, UseJitter, Options) ->
    case Fun() of
        {ok, Result} ->
            {ok, Result};
        {error, Reason} = Error ->
            %% Check if error is retryable
            Retryable =
                case maps:get(retryable_errors, Options, undefined) of
                    undefined ->
                        is_retryable_error(Error);
                    CustomCheck when is_function(CustomCheck, 1) ->
                        CustomCheck(Error);
                    _ ->
                        is_retryable_error(Error)
                end,

            case Retryable of
                true ->
                    %% Calculate backoff time with exponential increase
                    BaseBackoff = InitialBackoff * round(math:pow(Multiplier, Attempt)),
                    LimitedBackoff = min(BaseBackoff, MaxBackoff),

                    %% Add jitter if enabled (random 0-25% of backoff time)
                    FinalBackoff =
                        case UseJitter of
                            true ->
                                JitterAmount = rand:uniform(LimitedBackoff div 4),
                                LimitedBackoff + JitterAmount;
                            false ->
                                LimitedBackoff
                        end,

                    %% Log retry attempt if logger provided
                    case maps:get(logger, Options, undefined) of
                        undefined ->
                            ok;
                        Logger when is_function(Logger, 1) ->
                            Logger(#{
                                event => retry_attempt,
                                attempt => Attempt + 1,
                                max_retries => MaxRetries,
                                backoff_ms => FinalBackoff,
                                error => Reason
                            })
                    end,

                    %% Sleep before retry
                    timer:sleep(FinalBackoff),

                    %% Retry
                    retry_loop(
                        Fun,
                        Attempt + 1,
                        MaxRetries,
                        InitialBackoff,
                        MaxBackoff,
                        Multiplier,
                        UseJitter,
                        Options
                    );
                false ->
                    %% Non-retryable error - return immediately
                    {error, Reason}
            end
    end.
